packages/hash/frontend/src/blocks/page/BlockContextMenu/BlockContextMenu.tsx

```diff
@@ -0,0 +1,223 @@
+import { useUsers } from "../../../components/hooks/useUsers";
+import { FontAwesomeIcon } from "../../../shared/icons";
+import { BlockContextMenuItem } from "./BlockContextMenuItem";
+import { LoadEntityMenuContent } from "./LoadEntityMenuContent";
+import { BlockListMenuContent } from "./BlockListMenuContent";
+
+type BlockContextMenuProps = {
+  popupState: PopupState;
+  blockSuggesterProps: BlockSuggesterProps;
+  entityId: string | null;
+  entityStore: EntityStore;
+  view: EditorView<Schema>;
+};
+
+const LOAD_BLOCK_ENTITY_UI = "hash-load-entity-ui";
+
+export const BlockContextMenu = forwardRef<
+  HTMLDivElement,
+  BlockContextMenuProps
+>(({ popupState, blockSuggesterProps, entityId, entityStore }, ref) => {
+  const blockData = entityId ? entityStore.saved[entityId] : null;
+  const { data: users } = useUsers();
+  const setEntityMenuItemRef = useRef<HTMLLIElement>(null);
+  const swapBlocksMenuItemRef = useRef<HTMLLIElement>(null);
+
+  if (blockData && !isBlockEntity(blockData)) {
+    throw new Error("BlockContextMenu linked to non-block entity");
+  }
+
+  const menuItems = useMemo(() => {
+    const hasChildEntity =
+      Object.keys(blockData?.properties.entity?.properties ?? {}).length > 0;
+    const items = [
+      {
+        key: "set-entity",
+        title: hasChildEntity ? "Swap Entity" : "Add an entity",
+        icon: <FontAwesomeIcon icon={faAdd} />,
+        // @todo fix this TS issue. all subMenu's require
+        // popupState and that's passed in through cloneElement
+        // in BlockContextMenuItem
+        // For now the temporary fix is to make popupState optional
+        // in LoadEntityMenuContent
+        subMenu: <LoadEntityMenuContent entityId={entityId} />,
+        subMenuWidth: 280,
+      },
+      {
+        key: "copy-link",
+        title: "Copy Link",
+        icon: <FontAwesomeIcon icon={faLink} />,
+        onClick: () => {
+          const url = new URL(document.location.href);
+          url.hash = getBlockDomId(entityId!);
+          void navigator.clipboard.writeText(url.toString());
+        },
+      },
+      {
+        key: "duplicate",
+        title: "Duplicate",
+        icon: <FontAwesomeIcon icon={faCopy} />,
+      },
+      {
+        key: "delete",
+        title: "Delete",
+        icon: <FontAwesomeIcon icon={faTrashCan} />,
+      },
+      {
+        key: "swap-block",
+        title: "Swap block type",
+        icon: <FontAwesomeIcon icon={faRefresh} />,
+        subMenu: (
+          <BlockListMenuContent blockSuggesterProps={blockSuggesterProps} />
+        ),
+        subMenuWidth: 228,
+      },
+      {
+        key: "move-to-page",
+        title: "Move to page",
+        icon: <FontAwesomeIcon icon={faArrowRight} />,
+      },
+      {
+        key: "comment",
+        title: "Comment",
+        icon: <FontAwesomeIcon icon={faMessage} />,
+      },
+    ];
+
+    if (!localStorage.getItem(LOAD_BLOCK_ENTITY_UI)) {
+      items.shift();
+    }
```

Comment:
The fix was to update `entityVersionCreatedAt` of the BlockEntity after its child entity has been updated, but you mentioned the client doesn't need to update that since the api is meant to do that. The entityVersionCreatedAt returned by the api doesn't currently get updated. Intend leaving this till when that is fixed

Reply:
Can you link or create a task describing the issue please, in known issues?

Reply:
Noted. Planned addressing in this [task](https://app.asana.com/0/1201959586244685/1202106892392942)